Move filter UI into a toggle-able panel to improve experience on narrow viewports/containers

jorgefilipecosta opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/60696

Adds a button to toggle the filter visibility. The button also shows the number of filters applied on the corner.

## Screenshot
<img width="1443" alt="Screenshot 2024-07-05 at 17 54 22" src="https://github.com/WordPress/gutenberg/assets/11271197/e31b614a-c67a-41af-9f41-91d7905199c8">
<img width="1442" alt="Screenshot 2024-07-05 at 17 54 12" src="https://github.com/WordPress/gutenberg/assets/11271197/1dfba7f0-5724-4425-a7af-f9c422bf4f0f">



## Testing Instructions
I verified a button to toggle the filter visibility was available and worked as expected.
I verified that the count shown on the button matched the number of filters enabled and if there are no filters, the count is not shown.

jasmussen commented:
The button does suggest a drop-down. I also believe we had a dropdown design for filters somewhere, though there were likely complexity reasons for this. IMO this seems okay to me, as it improves the situation, but I'll defer to Jay or others!

jameskoster commented:
Thanks @jorgefilipecosta, this looks like a good start.

One quick change would be to ensure the panel is closed by default, rather than open. We might also try opening by default _when filters are applied_, but I don't feel as strongly about that part.

Given the visibility is now toggleable I think we can contain the filter UI within a dedicated row beneath `.dataviews-filters__view-actions` (we should probably rename that class too :D).

That structural change would decrease the likelihood of wrapping on narrow containers. In other words this:

<img width="409" alt="Screenshot 2024-07-12 at 12 13 25" src="https://github.com/user-attachments/assets/a106fbd3-83fb-482b-87de-cc7726cd841b">


Instead of this:

<img width="381" alt="Screenshot 2024-07-12 at 12 05 11" src="https://github.com/user-attachments/assets/70ed87f2-5a1c-4eff-9d6d-6cec7f3ebbe4">

jorgefilipecosta committed:
Remove isShowingFilter props, and use a key approach (includes some debug code). (+15 squashed commits)
Squashed commits:
[2198f16852] style change
[0f23576ae6] hide filters when changing views
[fbee4cad46] lint fix
[eaaccae54b] end to end test updates
[7b58faec1c] make filters toggle at the side of search
[2f5a957a47] end to end test updates
[9e40065f17] end to end test updare
[9fd5ca03b5] lint filters and make the filter button the first
[877f13bd54] Count styling
[4fc07717be] Reduce spacing, remove negative margin
[871ebf4553] Remove negative margin
[0d8632a666] lint fix
[31d893beb3] post rebase fixes
[1d2b4bf1d3] fix don't show all filters
[224d0ef7b1] fix number positioning (+3 squashed commits)
Squashed commits:
[f0acb40a7d] enhanements
[309567963c] post rebase style fixes
[9ce5f3686a] Remove hstack form search (+2 squashed commits)
Squashed commits:
[62c7124201] lint fixes
[afa224734a] apply feedback

ntsekouras commented:
It feels a bit weird that we have to click the toggle to then be presented with the `add filter` button, no? This is in the cases where no filter is active.

ntsekouras commented on packages/dataviews/src/dataviews.tsx:
```diff
@@ -150,7 +147,33 @@ export default function DataViews< Item >( {
+						isPressed={ isShowingFilter }
+						aria-expanded={ isShowingFilter }
+					/>
+					{ !! view.filters?.length && view.filters?.length > 0 && (
```
Nit: `!! view.filters?.length` should be enough.

ntsekouras commented:
There is one more nuance with `openedFilter`. This was added initially to in order to open the filter when we add it for the first time. With these changes, if we add a filter and then toggle the filter visibility, it still opens the last inserted filter.

jameskoster commented:
> It feels a bit weird that we have to click the toggle to then be presented with the add filter button, no? This is in the cases where no filter is active.

Yes, I think that's a good point. When no filters are set the filter button could expose the 'Add filter' menu instead. Here's the flow:

![add-filter](https://github.com/user-attachments/assets/9a1f8a98-7200-4b9e-bcc5-8e4617aa968a)

Once filters are added it behaves like a toggle for the panel.

jorgefilipecosta commented:
> There is one more nuance with `openedFilter`. This was added initially to in order to open the filter when we add it for the first time. With these changes, if we add a filter and then toggle the filter visibility, it still opens the last inserted filter.

Hi @ntsekouras, I think this issue is fixed.

> Yes, I think that's a good point. When no filters are set the filter button could expose the 'Add filter' menu instead. Here's the flow:

Hi @jameskoster, the solution you suggested was implemented, let me know if it looks as expected.

(the end to end tests are not updated yet, but if we agree on this solution I will do the update on this PR)

ntsekouras commented on packages/dataviews/src/dataviews.tsx:
```diff
@@ -111,7 +119,7 @@ export default function DataViews< Item >( {
 				<HStack
 					justify="start"
-					className="dataviews-filters__container"
+					className="dataviews-search__container"
```
I guess we won't need the `hstack` now.. We need to verify though if it has visual side effects..

jorgefilipecosta replied:
Yes, we can remove hstack but I needed to add a small, a small CSS rule to make the search still expand.
The search input also got a little bit bigger but I think we have space so it may be ok to default to the normal input control width (240px). Going to cc: @WordPress/gutenberg-design to check if we can do this small dimension change.

ntsekouras commented on packages/dataviews/src/filters.tsx:
```diff
@@ -22,6 +24,120 @@ interface FiltersProps< Item > {
+					<Button
+						size="compact"
+						icon={ funnel }
+						label={ __( 'Toggle filter display' ) }
```
Should this label be `add filter` in this case?

jorgefilipecosta replied:
Yes it should, it was updated 👍

ntsekouras commented:
This tests well and code wise looks good. Let's see the design input and see to update the tests afterwards.

jameskoster commented:
This is working really well!

A couple of tiny visual tweaks:

* The gap between the Layout and Settings icon buttons is 4px. But the gap between those buttons and the Filter button is 8px. Ideally that gap should be 4px too. 
* The toggle button is spilling out of its container breaking alignment with the primary action button above. `flex-shrink: 0` on `.dataviews-filters-toggle__container` seems to fix.
* On List layout we can remove the `max-width` applied to the search container so that it fills the available space.

jorgefilipecosta commented:
This PR passed by a big rebase, but it seems now everything is working again.

jorgefilipecosta commented:
Hi @jameskoster, thank you for the reviews, the first two tweaks you suggested were applied. The third one is being done on this PR https://github.com/WordPress/gutenberg/pull/63755 as it is a change to a behavior already on the trunk.

jameskoster commented:
Thanks for adding those tweaks :) I'm noticing a small bug; on the Pages table, adding one filter seems to add all filter options:

https://github.com/user-attachments/assets/b1fcf10b-598c-4bed-a230-849eb93b0c49

jorgefilipecosta commented:
> I'm noticing a small bug; on the Pages table, adding one filter seems to add all filter options:

Hi @jameskoster, thank you for catching this it was an issue introduced during a rebase and is now fixed.

jameskoster reviewed (commented):
I pushed a couple of small style tweaks, I think this is ready to go now, but let's see if some more feedback comes in :)

youknowriad commented:
Probably not specific to this branch, but why are we losing filters when switching the layout type (testing on pages dataviews)

youknowriad commented:
<img width="480" alt="Screenshot 2024-07-23 at 10 03 36 AM" src="https://github.com/user-attachments/assets/8fa876fb-fb5f-46f9-8d3c-784c9a97a54e">

I feel like there's something wrong here personally. Too much buttons and controls.

youknowriad commented:
I also think it's more logical for the filters toggle to be closer to the "search" input (left of the bar rather than right)

jorgefilipecosta commented:
> Probably not specific to this branch, but why are we losing filters when switching the layout type (testing on pages dataviews)

Good catch @youknowriad, that's not an issue of this branch. I will try to fix this in a separate PR.

jorgefilipecosta commented:
So should we try to reorder the items, making bulk edit the last one and the filters the first one (closer to the search)? Would that work @jameskoster, @youknowriad?

On https://github.com/WordPress/gutenberg/pull/63852 I am also moving the item size picker reducing the number of buttons/controls at the top.

jameskoster commented:
Yeah let's try moving the filter button closer to the search input.

Worth noting that the 'Bulk edit' button will also be moved (https://github.com/WordPress/gutenberg/issues/63193) which will help the the over-crowding.

jorgefilipecosta commented:
I made the filters button the first one as suggested.

youknowriad commented:
How do you feel about the button behaving sometimes as a dropdown menu (initial filter) and sometimes as a toggle when a filter is already applied. Should we be consistent and make it always a toggle?

youknowriad commented:
Also should we move it right after the search input (left area)?

jorgefilipecosta commented:
> How do you feel about the button behaving sometimes as a dropdown menu (initial filter) and sometimes as a toggle when a filter is already applied. Should we be consistent and make it always a toggle?

Hi @youknowriad, being always a toggle was the initial solution I proposed, but following feedback from @ntsekouras that it felt weird to need to show the filters and then press add filter https://github.com/WordPress/gutenberg/pull/63203#issuecomment-2230922691 we implemented this behavior. I think the behavior works well and is not uncommon, some applications use something similar.

youknowriad commented:
> Hi @youknowriad, being always a toggle was the initial solution I proposed, but following feedback from @ntsekouras that it felt weird to need to show the filters and then press add filter https://github.com/WordPress/gutenberg/pull/63203#issuecomment-2230922691 we implemented this behavior. I think the behavior works well and is not uncommon, some applications use something similar.

Ok, I guess we can try and see.

jameskoster commented:
@jorgefilipecosta I reckon we could move it right next to the search input with an 8px gap between:

<img width="317" alt="Screenshot 2024-07-24 at 13 31 09" src="https://github.com/user-attachments/assets/e25a0f07-c34a-4840-831c-100b6635cb70">

jorgefilipecosta commented:
Hi @jameskoster the filter toggle/add button is now at the side of the search as suggested.

jameskoster commented:
I pushed a tiny style tweak to the count. 

There's only one small issue I see outstanding: The filter panel visibility persists as you navigate between different views. IE if I open the panel on one view, it stays open when I navigate to a different view. Ideally each view treats the open/close state dependently. Would that be possible?

jorgefilipecosta commented:
Hi @jameskoster thank you for the style enhancement.

> There's only one small issue I see outstanding: The filter panel visibility persists as you navigate between different views. IE if I open the panel on one view, it stays open when I navigate to a different view. Ideally each view treats the open/close state dependently. Would that be possible?

I made an update, and now the state should not persist across the views.

youknowriad commented on packages/dataviews/src/components/dataviews/index.tsx:
```diff
@@ -49,6 +49,8 @@ type DataViewsProps< Item > = {
 	onChangeSelection?: ( items: string[] ) => void;
 	header?: ReactNode;
+	isShowingFilter?: boolean;
+	setIsShowingFilter?: React.Dispatch< React.SetStateAction< boolean > >;
```
I would love if avoid these props please. I don't think we should be adding random props like that to the DataViews component.

Why do we need them?

jorgefilipecosta replied:
Hi @youknowriad, we added these props to reset the isShowingFilter state when the active view changes (e.g: from drafts, to scheduled). This allows us to control or reset the state from outside of the component. If these props aren't passed, the component will use its own state, but having them allows outside control when necessary.

youknowriad replied:
I see. I think it's not ideal in terms of Api for the DataViews component. Why control this state and not other states like dropdown states... I think this doesn't make for a great API in DataViews.

Two options from me:

 - Can we accept that this remains an internal state.
 - If it's about resetting the state, we can force remount with a separate `key` on DataViews if needed.

jorgefilipecosta replied:
> If it's about resetting the state, we can force remount with a separate key on DataViews if needed.

I have thought of this option but I think it will probably also have undesired effects. I can give it a try and see how it goes.

jorgefilipecosta committed:
fix search invalidates the views

jorgefilipecosta commented:
Hi @youknowriad, with https://github.com/WordPress/gutenberg/pull/63889 merged and an additional small fix on search (included in this PR) the key-based solution to reset the filter showing state seems to be working. This PR is ready for review.

youknowriad reviewed (approved):
This is working well for me. What do you think @jameskoster

oandregal commented:
@jameskoster @jorgefilipecosta what are your thoughts regarding filters that declare to be primary?

- Before this PR, primary filters were always visible and not listed in the "Add filter" component.
- After this PR, they still are not listed in the "Add filter" component but they are hidden until you click on "Add filter".

There're some inconsistencies with regards to the +1 number as well. Note how it's not set in Patterns (sync is a "primary" filter) but it's set in Pages (all secondary filters) before the user picks a value:

https://github.com/user-attachments/assets/7470a704-3c2b-4a47-9320-fa28aac3ff70

So, I wonder if the concept of "primary filters" still make sense. If so, I believe we should make them visible as before. If we don't want to do that, I think we could just remove the `filter.isPrimary` API because the only thing it does now is to hide the filters from the "Add filter" component.

oandregal commented:
Also note that the "Add filter" button is hidden if all filters are primary, and that creates weird interactions. Compare Patterns (only 1 filter that is primary) vs Templates (only 1 filter but it is NOT primary). It's not clear why the "Add filter" button would be visible but disabled in the later but hidden in the former:

https://github.com/user-attachments/assets/99b6632a-ac60-4bdb-a981-1d2c52b87272

jameskoster commented:
> There're some inconsistencies with regards to the +1 number as well

We should fix that 👍 The question is whether to show the count when a filter is added, or when a value is selected. I can see arguments for both. What do you think?

The intent in the [original issue](https://github.com/WordPress/gutenberg/issues/60696) was for the panel to be open automatically if any filters are added. So if a view had primary filters the panel would be open. Looks like we missed tracking that as a follow-up.

> It's not clear why the "Add filter" button would be visible but disabled in the later but hidden in the former

I suspect this is a quirk relating the the a11y requirements. Most views supply non-primary filters that users can add as required. When all filters are added, the button remains accessible to help folks understand no more filters can be added, and to avoid confusion around the button vanishing from the UI.  Because there's only a single filter on the Patterns page, and it's primary, users cannot ever add a new filter here. Consequently it's okay to hide the button entirely.

oandregal commented:
Follow-up for having the filter toggle opened on load if there are primary filters https://github.com/WordPress/gutenberg/pull/64651

oandregal commented:
There's another follow-up: what to do with the filter toggle when there are no filters?  Report: https://github.com/WordPress/gutenberg/issues/64631

There's this PR that hides the toggle https://github.com/WordPress/gutenberg/pull/64640

Alternatively, we could show it as "disabled". However, I find the disabled state is useful when there's opportunities for the component to be active. In this case, given there are no filters, there're no opportunities for it to become active — hence my preference to just hide it. In the interest of velocity, I'll approve and go ahead with that PR. Let me know if you'd like to do something different.

jameskoster commented:
That seems reasonable to me 👍

oandregal commented:
There's also a couple of details about keyboard interaction / how the "add filter" button should work:

- https://github.com/WordPress/gutenberg/issues/64689
- https://github.com/WordPress/gutenberg/issues/64697
